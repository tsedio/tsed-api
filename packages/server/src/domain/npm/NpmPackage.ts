import {CollectionOf, Description, Email, Enum, Example, Required, Url} from "@tsed/schema";
import crypto from "crypto";

export enum NpmPackageType {
  OFFICIAL = "official",
  PREMIUM = "premium",
  THIRD_PARTY = "3rd-party"
}

export enum NpmPackageCategory {
  LOGGER = "logger",
  CLI = "cli",
  FRAMEWORK = "framework"
}

export class NpmPackageMaintainer {
  @Required()
  username: string;

  @Required()
  @Email()
  email: string;

  @Url()
  get avatar(): string {
    return `https://www.gravatar.com/avatar/${crypto.createHash("md5").update(this.email.toLowerCase()).digest("hex")}`;
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  set avatar(avatar: string) {}
}

export class NpmPackage {
  @Description("Npm package name")
  @Example("@tsed/common")
  name: string;

  @Description("Package description")
  description: string;

  @Description("Package version")
  version: string;

  @Description("Git repository")
  @Url()
  @Example("https://github.com/tsedio/tsed")
  repository: string;

  @Description("Npm url")
  @Url()
  @Example("https://www.npmjs.com/package/@tsed/common")
  npm: string;

  @Description("Package Icon")
  @Url()
  icon: string; // url or filename from /static/icons

  @Description("Home page to the repository")
  @Url()
  homepage: string;

  @Description("Bugs tracker url")
  @Url()
  bugs: string;

  @CollectionOf(NpmPackageMaintainer)
  maintainers: NpmPackageMaintainer[] = [];

  @Description("Npm downloads count")
  downloads = 0;

  @Description("Github stars count")
  stars = 0;

  @Description("Package tags")
  @CollectionOf(String)
  tags: string[] = [];

  constructor({
    name,
    description,
    version,
    repository,
    npm,
    icon,
    homepage,
    bugs,
    maintainers,
    downloads,
    stars,
    tags
  }: Partial<NpmPackage> = {}) {
    name && (this.name = name);
    description && (this.description = description);
    version && (this.version = version);
    repository && (this.repository = repository);
    npm && (this.npm = npm);
    icon && (this.icon = icon);
    homepage && (this.homepage = homepage);
    bugs && (this.bugs = bugs);
    maintainers && (this.maintainers = maintainers);
    downloads && (this.downloads = downloads);
    stars && (this.stars = stars);
    tags && (this.tags = tags);
  }

  @Description("Package type")
  @Enum(NpmPackageType)
  get type(): NpmPackageType {
    if (this.name.startsWith("@tsed/")) {
      return NpmPackageType.OFFICIAL;
    }

    if (this.name.startsWith("@tsedio/")) {
      return NpmPackageType.PREMIUM;
    }

    return NpmPackageType.THIRD_PARTY;
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  set type(type: NpmPackageType) {}

  @Description("Package category")
  @Enum(NpmPackageCategory)
  get category(): NpmPackageCategory {
    if (this.name.startsWith("tsed-cli-") || this.name.startsWith("@tsed/cli")) {
      return NpmPackageCategory.CLI;
    }

    if (this.name.startsWith("tsed-logger-") || this.name.startsWith("@tsed/logger")) {
      return NpmPackageCategory.LOGGER;
    }

    return NpmPackageCategory.FRAMEWORK;
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  set category(type: NpmPackageCategory) {}

  getRepositoryOwner() {
    if (this.repository?.match("github.com")) {
      const url = new URL(this.repository);
      const [, owner, repo] = url.pathname.replace(/.git$/, "").split("/");
      return {owner, repo};
    }

    return false;
  }
}
