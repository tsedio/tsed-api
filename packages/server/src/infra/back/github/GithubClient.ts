import {Octokit} from "@octokit/rest";
import {RequestParameters} from "@octokit/types";
import {InjectContext} from "@tsed/di";
import {PlatformContext} from "@tsed/common";
import {UseCache} from "@tsed/platform-cache";
import {Constant, Injectable} from "@tsed/di";
import {Unauthorized} from "@tsed/exceptions";
import {BaseLogClient} from "../http/BaseLogClient";

@Injectable()
export class GithubClient extends BaseLogClient {
  callee = "GITHUB";

  octokit: Octokit;

  @Constant("github.accessToken")
  token: string;

  @Constant("github.whitelist", [])
  whiteList: string[];

  @InjectContext()
  $ctx?: PlatformContext;

  get repos(): any {
    return this.octokit.repos;
  }

  $onInit() {
    this.octokit = new Octokit({
      auth: this.token
    });
    this.octokit.hook.after("request", async (response, options: RequestParameters) => {
      this.onSuccess({...response, ...options});
    });

    this.octokit.hook.error("request", async (error: Error, options: RequestParameters) => {
      this.onError(error, options);
      throw error;
    });
  }

  @UseCache({ttl: 3600 * 24 * 10, refreshThreshold: 900})
  async getInfo(owner: string, repo: string) {
    const {data} = await this.octokit.repos.get({owner, repo});
    return data;
  }

  @UseCache({ttl: 3600 * 24 * 10, refreshThreshold: 900})
  async getReleases(owner: string, repo: string, page: number, per_page: number) {
    const {data} = await this.octokit.repos.listReleases({owner, repo, page, per_page});
    return data;
  }

  @UseCache({ttl: 3600 * 24 * 10, refreshThreshold: 900})
  async getContributors(owner: string, repo: string, page: number, per_page: number) {
    const {data} = await this.octokit.repos.listContributors({owner, repo, page, per_page});
    return data;
  }

  checkWhitelist(repo: string) {
    if (!this.whiteList.includes(repo)) {
      throw new Unauthorized(`Unauthorized repository`);
    }
    return this;
  }
}
