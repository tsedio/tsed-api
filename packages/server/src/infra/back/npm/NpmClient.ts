import {PlatformContext, UseCache} from "@tsed/common";
import {Injectable, InjectContext} from "@tsed/di";
import {deserialize} from "@tsed/json-mapper";
import {Method} from "axios";
import url from "url";
import {NpmPackage} from "../../../domain/npm/NpmPackage";
import {HttpClient, HttpClientOptions} from "../http/HttpClient";

const REGEX_REGISTRY_ENFORCED_HTTPS = /^https?:\/\/([^\/]+\.)?(yarnpkg\.com|npmjs\.(org|com))(\/|$)/;
const REGEX_REGISTRY_PREFIX = /^(https?:)?\/\//i;
const REGEX_EXCLUDED_KEYWORDS = /hentai|porn/gi;

export function addSuffix(pattern: string, suffix: string): string {
  if (!pattern.endsWith(suffix)) {
    return pattern + suffix;
  }

  return pattern;
}

export const SCOPE_SEPARATOR = "%2f";

export interface NpmSearchResponse {
  objects: {package: any; score: any; searchScore: any}[];
}

export interface NpmRequestOptions extends HttpClientOptions {
  host?: string;
  unfiltered?: boolean;
  retry?: number;
}

@Injectable()
export class NpmClient extends HttpClient {
  hostRegistry = "https://registry.npmjs.org";
  hostApi = "https://api.npmjs.org";

  @InjectContext()
  context?: PlatformContext;

  static escapeName(name: string): string {
    // scoped packages contain slashes and the npm registry expects them to be escaped
    return name.replace("/", SCOPE_SEPARATOR);
  }

  getOptions(method: Method, endpoint: string, options: NpmRequestOptions): any {
    endpoint = this.getRequestUrl(options.host || this.hostRegistry, endpoint);
    options = super.getOptions(method, endpoint, options);

    return {
      ...options,
      headers: {
        Accept: options.unfiltered ? "application/json" : "application/vnd.npm.install-v1+json; q=1.0, application/json; q=0.8, */*",
        ...options.headers
      }
    };
  }

  getRequestUrl(registry: string, pathname: string): string {
    let resolved = pathname;

    if (!REGEX_REGISTRY_PREFIX.test(pathname)) {
      resolved = url.resolve(addSuffix(registry, "/"), pathname);
    }

    if (REGEX_REGISTRY_ENFORCED_HTTPS.test(resolved)) {
      resolved = resolved.replace(/^http:\/\//, "https://");
    }

    return resolved;
  }

  /**
   * Search a module on npm registry
   * @param text
   * @param options
   */
  @UseCache({
    ttl: 3600 * 24 * 10,
    refreshThreshold: 900,
    type: NpmPackage,
    collectionType: Array,
    key([type]: string[]): string {
      return `npm:search:${type}`;
    }
  })
  async search(
    text: string,
    options: {size?: number; from?: number; quality?: number; popularity?: number; maintenance?: number} = {}
  ): Promise<NpmPackage[]> {
    const {objects: result} = await this.get<NpmSearchResponse>(`-/v1/search`, {
      headers: {
        "Accept-Encoding": "gzip"
      },
      params: {
        text,
        size: 100,
        from: 0,
        quality: 0.65,
        popularity: 0.98,
        maintenance: 0.5,
        ...options
      }
    });

    const promises = result
      .filter(({package: obj}) => !obj.name.match(REGEX_EXCLUDED_KEYWORDS))
      .map<Promise<NpmPackage>>(async ({package: {links, ...props}}) => {
        const downloads = await this.downloads(props.name);

        return deserialize(
          {
            ...props,
            repository: links.repository,
            homepage: links.homepage,
            npm: links.npm,
            bugs: links.bugs,
            downloads
          },
          {type: NpmPackage}
        );
      })
      .filter(Boolean);

    return await Promise.all(promises);
  }

  async downloads(pkg: string): Promise<number> {
    try {
      const {downloads} = await this.get<{downloads: number}>(`/downloads/point/last-month/${pkg}`, {
        host: this.hostApi
      });

      return downloads;
    } catch (er) {
      this.context?.logger.warn({message: "Unable to get downloads for the following packages", pkg, error: er});
      return 0;
    }
  }
}
