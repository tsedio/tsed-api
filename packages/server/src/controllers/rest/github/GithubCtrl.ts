import {InjectContext} from "@tsed/di";
import {BodyParams, Controller, Get, HeaderParams, Inject, PathParams, PlatformContext, Post, QueryParams} from "@tsed/common";
import {array, boolean, Default, Max, Min, Name, number, object, Returns, string} from "@tsed/schema";
import {GithubWebhookPayload} from "../../../domain/github/GithubWebhookPayload";
import {GithubClient} from "../../../infra/back/github/GithubClient";
import {CacheService} from "../../../infra/persistence/CacheService";

const GithubRepo = object({
  id: string().description("Github repo id"),
  html_url: string().required(),
  stargazers_count: string().required().description("Github stargazers"),
  watchers_count: string().required().description("Github watchers"),
  forks_count: string().required().description("Github forks"),
  open_issues_count: string().required().description("Github open issues")
}).label("GithubRepo");

const GithubContributor = object({
  login: string().required(),
  id: string().required(),
  node_id: string().required(),
  avatar_url: string().required(),
  gravatar_id: string().required(),
  url: string().required(),
  html_url: string().required(),
  followers_url: string().required(),
  following_url: string().required(),
  gists_url: string().required(),
  starred_url: string().required(),
  subscriptions_url: string().required(),
  organizations_url: string().required(),
  repos_url: string().required(),
  events_url: string().required(),
  received_events_url: string().required(),
  type: string().required(),
  site_admin: string().required(),
  contributions: number().required()
}).label("GithubContributor");

const GithubContributors = array().items(GithubContributor).label("GithubContributors");

const GithubRelease = object({
  url: string().required().example("https://api.github.com/repos/tsedio/tsed/releases/38791984"),
  assets_url: string().required().example("https://api.github.com/repos/tsedio/tsed/releases/38791984/assets"),
  upload_url: string().required().example("https://uploads.github.com/repos/tsedio/tsed/releases/38791984/assets{?name,label}"),
  html_url: string().required().example("https://github.com/tsedio/tsed/releases/tag/v6.26.4"),
  id: number().required().example(38791984),
  author: GithubContributor,
  node_id: string().required().example("MDc6UmVsZWFzZTM4NzkxOTg0"),
  tag_name: string().required().example("v6.26.4"),
  target_commitish: string().required().example("production"),
  name: string().required().example("v6.26.4"),
  draft: boolean().required(),
  prerelease: boolean().required(),
  created_at: string().required().example("2021-02-26T10:35:52Z"),
  published_at: string().required().example("2021-02-26T10:36:31Z"),
  assets: array().items(object()),
  tarball_url: string().required().example("https://api.github.com/repos/tsedio/tsed/tarball/v6.26.4"),
  zipball_url: string().required().example("https://api.github.com/repos/tsedio/tsed/zipball/v6.26.4"),
  body: string().required()
}).label("GithubRelease");

const GithubReleases = array().items(GithubRelease).label("GithubReleases");

@Controller("/github/:owner/:repo")
@Name("Github")
export class GithubCtrl {
  @InjectContext()
  $ctx?: PlatformContext;

  @Inject()
  cache: CacheService;

  @Inject()
  protected client: GithubClient;

  handleResponse({data}: any) {
    return data;
  }

  @Get()
  @Returns(200).ContentType("application/json").Schema(GithubRepo)
  @Returns(401).Description("Repository unauthorized")
  async get(@PathParams("owner") owner: string, @PathParams("repo") repo: string) {
    const data = await this.client.checkWhitelist(repo).getInfo(owner, repo);

    return {
      id: data.id,
      html_url: data.html_url,
      stargazers_count: data.stargazers_count,
      watchers_count: data.watchers_count,
      forks_count: data.forks_count,
      open_issues_count: data.open_issues_count
    };
  }

  @Get("/contributors")
  @Returns(200).ContentType("application/json").Schema(GithubContributors)
  @Returns(401).Description("Repository unauthorized")
  async getContributors(@PathParams("owner") owner: string, @PathParams("repo") repo: string) {
    return this.client.checkWhitelist(repo).getContributors(owner, repo, 1, 100);
  }

  @Get("/releases")
  @Returns(200).ContentType("application/json").Schema(GithubReleases)
  @Returns(401).Description("Repository unauthorized")
  async getReleases(
    @PathParams("owner") owner: string,
    @PathParams("repo") repo: string,
    @QueryParams("page", Number) @Min(1) @Default(1) page = 1,
    @QueryParams("per_page", Number) @Max(100) @Default(30) per_page = 30
  ) {
    return this.client.checkWhitelist(repo).getReleases(owner, repo, page, per_page);
  }

  @Post("/webhook")
  @Returns(200, String).Examples("OK")
  async webhook(@HeaderParams("x-github-event") event: string, @BodyParams() payload: GithubWebhookPayload) {
    const [owner, repo] = payload.repository.fullName.split("/");

    switch (event) {
      case "release":
        await this.cache.deleteMatchingKeys(new RegExp(`GithubClient:getReleases:${owner}:${repo}`));
        await this.cache.deleteMatchingKeys(new RegExp(`GithubClient:getContributors:${owner}:${repo}`));
        break;
      case "star":
        await this.cache.deleteMatchingKeys(new RegExp(`GithubClient:getInfo:${owner}:${repo}`));
        break;
    }

    return "OK";
  }
}
