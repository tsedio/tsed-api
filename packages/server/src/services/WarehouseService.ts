import {InjectContext} from "@tsed/async-hook-context";
import {Injectable, PlatformContext, UseCache} from "@tsed/common";
import {toMap} from "@tsed/core";
import {Inject} from "@tsed/di";
import {FormioSubmission} from "@tsed/formio";
import {NpmPackage, NpmPackageType} from "../domain/npm/NpmPackage";
import {GithubClient} from "../infra/back/github/GithubClient";
import {NpmClient} from "../infra/back/npm/NpmClient";
import {FormioRepository} from "./FormioRepository";

export type SubmissionPackage = FormioSubmission<NpmPackage & {disabled: boolean}>;

const RANKS = {
  [NpmPackageType.PREMIUM]: 1,
  [NpmPackageType.OFFICIAL]: 0,
  [NpmPackageType.THIRD_PARTY]: 0
};

@Injectable()
export class WarehouseService extends FormioRepository {
  protected formName = "packages";

  @InjectContext()
  protected $ctx: PlatformContext;

  @Inject()
  protected npmClient: NpmClient;

  @Inject()
  protected githubClient: GithubClient;

  @UseCache({
    ttl: 3600 * 24 * 10,
    refreshThreshold: 900,
    type: NpmPackage,
    collectionType: Array
  })
  async getPlugins(keyword: string): Promise<NpmPackage[]> {
    const [packages, submissions] = await Promise.all([this.npmClient.search(keyword), this.getPackagesSubmissions()]);

    const submissionsPackagesMap = toMap(submissions, (item: SubmissionPackage) => item.data.name);

    const result = await Promise.all(
      packages.map(async (pkg) => {
        if (!submissionsPackagesMap.has(pkg.name)) {
          submissionsPackagesMap.set(pkg.name, await this.createPackage(pkg));
        }

        const submission = submissionsPackagesMap.get(pkg.name);
        submissionsPackagesMap.delete(pkg.name);

        if (submission.data.disabled) {
          return false;
        }

        pkg.icon = submission.data.icon;
        pkg.tags = submission.data.tags;
        pkg.description = submission.data.description || pkg.description;
        pkg.homepage = (submission.data.homepage || pkg.homepage || pkg.repository || pkg.npm || "").replace("//packages", "/packages");
        pkg.stars = await this.getStars(pkg);

        if (submission.data.maintainers) {
          pkg.maintainers.push(...submission.data.maintainers);
        }

        return pkg;
      })
    );

    const otherPackages = [...submissionsPackagesMap.values()].map((submissionPackage) => {
      return new NpmPackage({
        ...submissionPackage.data
      });
    });

    const allPackages = [...otherPackages, ...(result.filter(Boolean) as NpmPackage[])].filter(Boolean);

    return [
      ...allPackages.filter((pkg) => pkg.type === NpmPackageType.PREMIUM).sort(),
      ...allPackages
        .filter((pkg) => pkg.type != NpmPackageType.PREMIUM)
        .sort((p1, p2) => {
          return p1.downloads < p2.downloads ? 1 : -1;
        })
    ];
  }

  async getStars(pkg: NpmPackage) {
    const meta = pkg.getRepositoryOwner();

    if (!meta) {
      return 0;
    }

    try {
      const {stargazers_count} = await this.githubClient.getInfo(meta.owner, meta.repo);

      return stargazers_count;
    } catch (er) {
      this.$ctx.logger.error({
        event: "GITHUB_REPO_INFO_ERROR",
        error: er,
        owner: meta.owner,
        repo: meta.repo
      });
      return 0;
    }
  }

  getPackagesSubmissions(): Promise<SubmissionPackage> {
    return this.getSubmissions() as Promise<any>;
  }

  async createPackage(pkg: NpmPackage): Promise<SubmissionPackage> {
    return this.saveSubmission({
      data: {
        name: pkg.name,
        description: pkg.description,
        icon: ""
      }
    });
  }
}
