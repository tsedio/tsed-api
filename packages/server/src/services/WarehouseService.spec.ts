import {PlatformTest} from "@tsed/common";
import {deserialize} from "@tsed/json-mapper";
import {NpmPackage} from "../domain/npm/NpmPackage";
import {GithubClient} from "../infra/back/github/GithubClient";
import {NpmClient} from "../infra/back/npm/NpmClient";
import {WarehouseService} from "./WarehouseService";

async function getWarehouseFixture({data, npmData}: any) {
  const npmClient = {
    search: jest.fn().mockResolvedValue([
      deserialize(
        npmData || {
          name: "@tsed/common"
        },
        {type: NpmPackage}
      )
    ])
  };

  const service = await PlatformTest.invoke<WarehouseService>(WarehouseService, [
    {
      token: NpmClient,
      use: npmClient
    }
  ]);

  jest.spyOn(service, "getSubmissions").mockResolvedValue([data as any]);
  jest.spyOn(service, "saveSubmission").mockImplementation((o: any) => o);
  jest.spyOn(service, "getStars").mockResolvedValue(1000);

  return {npmClient, service};
}

describe("WarehouseService", () => {
  beforeEach(() => PlatformTest.create());
  afterEach(() => PlatformTest.reset());

  describe("getPlugins()", () => {
    it("should return all plugins available for the given keywords", async () => {
      const {service} = await getWarehouseFixture({
        data: {
          _id: "id",
          form: "formId",
          data: {
            name: "@tsed/common",
            icon: "",
            disabled: false,
            maintainers: [
              {
                username: "Romakita",
                email: "romakita@tsed.io"
              }
            ]
          }
        }
      });

      const result = await service.getPlugins("tsed");

      expect(result).toEqual([
        {
          downloads: 0,
          homepage: "",
          icon: "",
          maintainers: [
            {
              email: "romakita@tsed.io",
              username: "Romakita"
            }
          ],
          name: "@tsed/common",
          stars: 1000
        }
      ]);
    });
    it("should save the unknown package to database", async () => {
      const {service} = await getWarehouseFixture({
        data: {
          _id: "id",
          form: "formId",
          data: {
            name: "@tsed/common",
            icon: "",
            disabled: false
          }
        },
        npmData: {
          name: "@tsed/prisma",
          description: "A prisma package",
          version: "1.0.0",
          repository: "https://github.com/tsedio/tsed-prisma",
          homepage: "https://github.com/tsedio/tsed-prisma"
        }
      });

      const result = await service.getPlugins("tsed");

      expect(result).toEqual([
        {
          downloads: 0,
          maintainers: [],
          name: "@tsed/common",
          stars: 0,
          tags: []
        },
        {
          description: "A prisma package",
          downloads: 0,
          homepage: "https://github.com/tsedio/tsed-prisma",
          icon: "",
          maintainers: [],
          name: "@tsed/prisma",
          repository: "https://github.com/tsedio/tsed-prisma",
          stars: 1000,
          version: "1.0.0"
        }
      ]);
      expect(service.saveSubmission).toHaveBeenCalledWith({
        data: {
          description: "A prisma package",
          icon: "",
          name: "@tsed/prisma"
        }
      });
    });
    it("should return nothing when the package is mentioned as disabled", async () => {
      const {service} = await getWarehouseFixture({
        data: {
          _id: "id",
          form: "formId",
          data: {
            name: "@tsed/common",
            icon: "",
            disabled: true
          }
        }
      });

      const result = await service.getPlugins("tsed");

      expect(result).toEqual([]);
    });
    it("should get icon from formio", async () => {
      const {service} = await getWarehouseFixture({
        data: {
          _id: "id",
          form: "formId",
          data: {
            name: "@tsed/common",
            icon: "https://icon.formio",
            disabled: false
          }
        }
      });

      jest.spyOn(service, "getStars").mockResolvedValue(1000);

      const result = await service.getPlugins("tsed");

      expect(result).toEqual([
        {
          downloads: 0,
          homepage: "",
          icon: "https://icon.formio",
          maintainers: [],
          name: "@tsed/common",
          stars: 1000
        }
      ]);
    });
  });

  describe("getStars()", () => {
    it("should get stargazers_count from github", async () => {
      const githubClient = {
        getInfo: jest.fn().mockResolvedValue({stargazers_count: 1000})
      };
      const service = await PlatformTest.invoke<WarehouseService>(WarehouseService, [
        {
          token: GithubClient,
          use: githubClient
        }
      ]);

      const pkg = new NpmPackage();
      pkg.name = "@tsed/common";
      pkg.repository = "https://github.com/tsedio/tsed";

      const result = await service.getStars(pkg);

      expect(result).toEqual(1000);
      expect(githubClient.getInfo).toHaveBeenCalledWith("tsedio", "tsed");
    });

    it("should return 0 when the repository isn't a github", async () => {
      const githubClient = {
        getInfo: jest.fn().mockResolvedValue({stargazers_count: 1000})
      };
      const service = await PlatformTest.invoke<WarehouseService>(WarehouseService, [
        {
          token: GithubClient,
          use: githubClient
        }
      ]);

      const pkg = new NpmPackage();
      pkg.name = "@tsed/common";
      pkg.repository = "https://gitlab.com/tsedio/tsed";

      const result = await service.getStars(pkg);

      expect(result).toEqual(0);
      expect(githubClient.getInfo).not.toHaveBeenCalled();
    });
  });
});
