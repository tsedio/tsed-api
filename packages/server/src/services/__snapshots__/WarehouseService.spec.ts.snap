// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarehouseService getPlugins() should save the unknown package to database 1`] = `
Array [
  NpmPackage {
    "description": "A prisma package",
    "downloads": 0,
    "homepage": "https://github.com/tsedio/tsed-prisma",
    "icon": "",
    "maintainers": Array [],
    "name": "@tsed/prisma",
    "repository": "https://github.com/tsedio/tsed-prisma",
    "stars": 1000,
    "tags": undefined,
    "version": "1.0.0",
  },
  NpmPackage {
    "downloads": 0,
    "maintainers": Array [],
    "name": "@tsed/common",
    "stars": 0,
    "tags": Array [],
  },
]
`;
